npm/src/upload/upload-form.ts

```diff
@@ -63,83 +135,117 @@ export class UploadForm {
-      }
+  addFolderListener() {
+    this.dropzone.addEventListener("drop", this.handleDropppedItems)
+
+    this.folderRetriever.addEventListener("change", () => {
+      const form: HTMLFormElement | null = this.formElement
+      const files = this.retrieveFiles(form)
+      this.selectedFiles = files
+      const parentFolder = this.getParentFolderName(this.selectedFiles)
+      this.updateFolderSelectionStatus(parentFolder, files.length)
     })
   }
 
   addSubmitListener(
-    uploadFiles: (files: TdrFile[], uploadFilesInfo: FileUploadInfo) => void
+    uploadFiles: (
+      files: IFileWithPath[],
+      uploadFilesInfo: FileUploadInfo
+    ) => void
   ) {
     this.formElement.addEventListener("submit", (ev) => {
       ev.preventDefault()
-      const target: HTMLInputTarget | null = ev.currentTarget
-      const files = this.retrieveFiles(target)
-      const parentFolder = this.getParentFolderName(files)
+      const parentFolder = this.getParentFolderName(this.selectedFiles)
       const uploadFilesInfo: FileUploadInfo = {
         consignmentId: this.consignmentId(),
         parentFolder: parentFolder
       }
-      uploadFiles(files, uploadFilesInfo)
+      uploadFiles(this.selectedFiles, uploadFilesInfo)
     })
   }
 
-  private retrieveFiles(target: HTMLInputTarget | null): TdrFile[] {
-    const files: TdrFile[] = target!.files!.files!
+  private getParentFolderName(folder: IFileWithPath[]) {
+    const firstItem: FileWithRelativePath = folder[0]
+      .file as FileWithRelativePath
+    const relativePath: string = firstItem.webkitRelativePath
+    const splitPath: string[] = relativePath.split("/")
+    const parentFolder: string = splitPath[0]
+    return parentFolder
+  }
+
+  private checkIfFolderHasFiles(files: File[] | IFileWithPath[]): void {
     if (files === null || files.length === 0) {
-      this.dropzone.classList.remove("drag-and-drop__dropzone--dragover")
-      const successMessage: HTMLElement | null = document.querySelector(
-        ".drag-and-drop__success"
-      )
-      successMessage?.classList.add("hide")
+      this.rejectUserItemSelection()
+    }
+  }
+
+  private retrieveFiles(target: HTMLInputTarget | null): IFileWithPath[] {
+    const files: File[] = target!.files!.files!
+    this.checkIfFolderHasFiles(files)
+
+    return [...files].map((file) => ({
+      file,
+      path: (file as FileWithRelativePath).webkitRelativePath
+    }))
+  }
+
+  private rejectUserItemSelection() {
+    this.dropzone.classList.remove("drag-and-drop__dropzone--dragover") // Extract this into its own method to do with file selection
+    const successMessage: HTMLElement | null = document.querySelector(
+      ".drag-and-drop__success"
+    )
+    successMessage?.classList.add("hide")
+    const warningMessage: HTMLElement | null = document.querySelector(
+      ".drag-and-drop__failure"
+    )
+    warningMessage?.classList.remove("hide")
+    throw new Error("No files selected")
```

Comment:
Do we need to throw an exception here? If we display the warning message like we do just above here, do we also need to throw an error?

Reply:
Hmm, it was there from before, it did serve me well when debugging (if you have multiple red warning messages in a row e.g. you upload multiple folders one time and then you upload a file), you might not realise that there were two errors) but it's probably not necessary now.

Reply:
There's few places in the code this is used. It is possible to refactor it with if else statements but I just tried and it gets messy quite quickly. It's just a convenient way of stopping the execution of the code that follows.